### DISCUSSION TITLE
How to measure end effector force?

### DISCUSSION THREAD

**Discussion by BolunDai0216** [bulletphysics/bullet3#3856] [Q&A]
Hi,

I have a manipulator and I want to measure the force applied to the environment, would anyone know how to measure that?

Thanks!

**Comment by erwincoumans** [comment]
There are a few ways: 
1) Use the contact points, and read the normal (and friction) forces.
2) Use enableJointForceTorqueSensor to enable joint feedback on the last link (having a fixed link helps, for 6-dof feedback), read the data using getJointState.

Having an example how to do this would help indeed.

See the PyBullet Quickstart Guide for more details.
https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.2ye70wns7io3

**Reply by BolunDai0216** [reply]
Thanks for the reply!

For enableJointForceTorqueSensor with a fixed link as the end effector will it just return the external force or will it also give the force and torque applied from the rest of the links?

An example would be greatly appreciated!

**Reply by BolunDai0216** [reply]
I tried using the contact points, however, the results are different from what I imagined. I used changeDynamics to set the stiffness of the object I want to come into contact as k and damping to 0. At each contact point there is a contactDistance value, should the normalForce be the same as k* contactDistance?

**Reply by erwincoumans** [reply]
enableJointForceTorqueSensor will report all the forces acting on the joint (assuming a fixed joint).

The contact forces at the contact points should work just fine. It is not the same as k * contactDistance.

**Reply by BolunDai0216** [reply]
> enableJointForceTorqueSensor will report all the forces acting on the joint (assuming a fixed joint).
> 
> The contact forces at the contact points should work just fine. It is not the same as k * contactDistance.

Is it possible to point me to some references to better understand how the normal force is calculated?
